design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,592 @@
+  so a user should be able to observe all runtime actions.
+
+Nice to have ideas that are out-of-scope of the first implementation of the actor model for thin-edge:
+* It would be a plus, to have actors storing data using the framework:
+  - to persist data between restarts of the deployment
+  - to cache data during network outages
+  - to provide operation checkpoints during sensitive operations.
+* It would be handy to build batteries-included executables
+  that contains numerous extensions and alternative implementations for a large diversity of use-cases.
+  - An extension can be included in an executable without being enabled, but only registered and ready to be used.
+  - Registered extensions can be enabled and configured at runtime. 
+  - Having extensions included but not instantiated should not impact other extensions nor consume runtime resources.
+  - Such an executable with enabled and non-enabled extensions must provide a command line option, say `--list-extensions`,
+    to provide the whole list of available extensions and their purpose;
+    as well as command line options for detailed help on how to enable and configure those.
+
+## Detailed design aspects
+
+The point here is not to detail all the design aspects, but to address the points that are important and challenging.
+
+### Actor sketch
+
+The design key ideas exposed above aim to let the actor implementations be as unconstrained as possible.
+
+Message types can be freely defined. They only need to be free of concurrency and ownership constraints.
+
+```rust
+/// A message has simply to be Debug + Send + Sync + 'static
+#[derive(Debug)]
+pub struct MyInput {
+  /// message fields must be Send + Sync + 'static
+}
+```
+
+All the interactions of an actor with peers are done via `mpsc` channels.
+An actor implementation has to own a set of `Receiver` and `Sender` along its state.
+
+```rust
+/// A typical actor owns a single receiver and a bunch of senders.
+/// 
+/// But this is not a rule.
+/// An actor can have single sender for all its output.
+/// Or `tokio::select!` over a bunch of receivers.
+pub struct MyActor {
+    /// The state that rules the actor behavior
+    state: MyActorSate,
+    
+    /// The queue of input messages for this actor
+    input_box: Receiver<MyInput>,
+    
+    /// Different output channels targeting specific peers
+    peer_a: Sender<PeerAInput>,
+    peer_b: Sender<PeerBInput>,
+    peer_c: Sender<PeerCInput>,
+}
+```
+
+The behavior of an actor is simply defined by an `async fn run(&mut self)` method.
+Note that no extra context, runtime or peer information are passed to the actor event loop.
+
+```rust
+impl MyActor {
+  /// The actor event loop
+  pub async fn run(&mut self) -> Result<(), RuntimeError> {
+    // - consume messages received on `self.input_box`
+    // - send messages to peers over `self.peer_XXX`
+    // - update `self.state`
+    // - forward errors to the appropriate peer 
```

Comment:
```suggestion
    // - consume messages received on `self.input_box`
    // - update `self.state`
    // - send messages/errors to peers over `self.peer_x` senders
```
Just to align with the typical flow: receive input -> process -> send output/error

Reply:
Updated